rotors_gazebo_plugins/src/gazebo_mavlink_interface.cpp

```diff
@@ -466,332 +273,319 @@ void GazeboMavlinkInterface::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf
+    // 11km above MSL)
+    const float lapse_rate =
+        0.0065f;  // reduction in temperature with altitude (Kelvin/m)
+    const float temperature_msl = 288.0f;  // temperature at MSL (Kelvin)
+    float alt_msl = (float)alt_home - pos_n.Z();
+    float temperature_local = temperature_msl - lapse_rate * alt_msl;
+    float pressure_ratio = powf((temperature_msl / temperature_local), 5.256f);
+    const float pressure_msl = 101325.0f;  // pressure at MSL
+    sensor_msg.abs_pressure = pressure_msl / pressure_ratio;
+
+    // generate Gaussian noise sequence using polar form of Box-Muller
+    // transformation http://www.design.caltech.edu/erik/Misc/Gaussian.html
+    double x1, x2, w, y1, y2;
+    do {
+      x1 = 2.0 * (rand() * (1.0 / (double)RAND_MAX)) - 1.0;
+      x2 = 2.0 * (rand() * (1.0 / (double)RAND_MAX)) - 1.0;
+      w = x1 * x1 + x2 * x2;
+    } while (w >= 1.0);
+    w = sqrt((-2.0 * log(w)) / w);
+    y1 = x1 * w;
+    y2 = x2 * w;
+
+    // Apply 1 Pa RMS noise
+    float abs_pressure_noise = 1.0f * (float)w;
+    sensor_msg.abs_pressure += abs_pressure_noise;
+
+    // convert to hPa
+    sensor_msg.abs_pressure *= 0.01f;
+
+    // calculate density using an ISA model for the tropsphere (valid up to 11km
+    // above MSL)
+    const float density_ratio =
+        powf((temperature_msl / temperature_local), 4.256f);
+    float rho = 1.225f / density_ratio;
+
+    // calculate pressure altitude including effect of pressure noise
+    sensor_msg.pressure_alt =
+        alt_msl - abs_pressure_noise / (gravity_W_.Length() * rho);
+
+    // calculate differential pressure in hPa
+    sensor_msg.diff_pressure = 0.005f * rho * vel_b.X() * vel_b.X();
+
+    // calculate temperature in Celsius
+    sensor_msg.temperature = temperature_local - 273.0f;
+
+    sensor_msg.fields_updated = 4095;
+
+    // accumulate gyro measurements that are needed for the optical flow message
+    static uint32_t last_dt_us = sensor_msg.time_usec;
+    uint32_t dt_us = sensor_msg.time_usec - last_dt_us;
+    if (dt_us > 1000) {
+      optflow_gyro += gyro_b * (dt_us / 1000000.0f);
+      last_dt_us = sensor_msg.time_usec;
+    }
 
-  send_mavlink_message(MAVLINK_MSG_ID_HIL_SENSOR, &sensor_msg, 200);
+    mavlink_message_t msg;
+    mavlink_msg_hil_sensor_encode_chan(
+        1, 200, MAVLINK_COMM_0, &msg, &sensor_msg);
+    if (hil_mode_) {
+      if (!hil_state_level_) {
+        send_mavlink_message(&msg);
+      }
+    }
+
+    else {
+      send_mavlink_message(&msg);
+    }
+    last_imu_time_ = current_time;
+  }
 
   // ground truth
-  ignition::math::Vector3d accel_true_b = q_br.RotateVector(model_->RelativeLinearAccel());
+#if GAZEBO_MAJOR_VERSION >= 9
+  ignition::math::Vector3d accel_true_b =
+      q_br.RotateVector(model_->RelativeLinearAccel());
+#else
+  ignition::math::Vector3d accel_true_b = q_br.RotateVector(
+      ignitionFromGazeboMath(model_->GetRelativeLinearAccel()));
```

Comment:
Not sure if ignitionFromGazeboMath is needed. If not, this is handled by autobackport. Otherwise we can leave it like that (it only one line..)